pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1218,7 +1223,22 @@ public void run(Timeout timeout) throws Exception {
     };
 
     @Override
-    protected void flush() {
+    public CompletableFuture<Void> flushAsync() {
+        if (isBatchMessagingEnabled()) {
+            CompletableFuture<MessageId> lastSendFuture;
+            synchronized (ProducerImpl.this) {
+                batchMessageAndSend();
```

Comment:
batchMessageAndSend is a no-op is batching is disabled, so this whole method could be just
```
public CompletableFuture<Void> flushAsync() {
    synchronized (ProducerImpl.this) {
        batchMessageAndSend();
        return lastSendFuture.thenApply(ignored -> null);
    }
}
```